### ISSUE TITLE
Generated .zips can't be opened by Kodi under Windows 10

### ISSUE THREAD

**Issue by zim514** [chadparry/kodi-repository.chad.parry.org#10]
I have a repo for dev versions of my addon hosted on github, but the .zip files generated by the script can't be loaded by Kodi.

You can see the repo source here: https://github.com/zim514/zim514.github.io/tree/master/repo . I'm using the latest version of the script under windows with Python 3.9.1, with Kodi 19.1. I've tried both the official latest windows and CoreElec versions of Kodi.

If I manually .zip the addons with 7zip, they work just fine so I assume my addon.xml is ok.  However, it appears that the .zip files generated by the script have a problem. They seem to download but consistently give an error that they can't be read. The generated .zip files open fine in 7zip, but not Kodi. An example broken zip is located here.

A full example log is here: [oqizuxaxov](https://paste.kodi.tv/oqizuxaxov) (paste)

Another forum user has tried as well and [had the same results](https://forum.kodi.tv/showthread.php?tid=363678&pid=3051087#pid3051087), though the script works fine for his repo.

Excerpt: 

```
2021-07-23 14:11:21.512 T:13924 ERROR <general>: CAddonInfoBuilder::ADDON::CAddonInfoBuilder::Generate: Unable to load 'zip://special%3a%2f%2fhome%2faddons%2fpackages%2fscript.service.hue-1.0.2.zip/script.service.hue/addon.xml', Line 0 Failed to open file
2021-07-23 14:11:21.512 T:13924 ERROR <general>: CAddonInstallJob[script.service.hue]: invalid package special://home/addons/packages/script.service.hue-1.0.2.zip
```

**Comment by zim514** [comment]
This patch fixes the issue under windows: [https://gist.githubusercontent.com/asavah/f1dbd90a839fd77faffe4a354b618940/raw/45fbffdcf3b1e52ca0eab8f737aa4c9899859a54/create_repository_win.patch](https://gist.githubusercontent.com/asavah/f1dbd90a839fd77faffe4a354b618940/raw/45fbffdcf3b1e52ca0eab8f737aa4c9899859a54/create_repository_win.patch
)

**Comment by zim514** [comment]
Thanks to forum users the problem was found and fixed.

It looks like the generated zip is missing the directory entry:

```
zipinfo linux-repository.snapcase-1.0.8.zip
Archive:  linux-repository.snapcase-1.0.8.zip
Zip file size: 125267 bytes, number of entries: 4
drwx---     0.0 fat        0 bx stor 21-Jul-24 22:01 repository.snapcase/
-rw----     0.0 fat     1156 tx defN 21-Jul-24 22:01 repository.snapcase/addon.xml
-rw----     0.0 fat   116361 bx defN 21-Jul-24 22:01 repository.snapcase/fanart.jpg
-rw----     0.0 fat    12978 bx defN 21-Jul-24 22:01 repository.snapcase/icon.png
4 files, 130495 bytes uncompressed, 124655 bytes compressed:  4.5%

zipinfo win-repository.snapcase-1.0.8.zip
Archive:  win-repository.snapcase-1.0.8.zip
Zip file size: 125137 bytes, number of entries: 3
-rw----     0.0 fat     1179 tx defN 21-Jul-24 22:20 repository.snapcase\addon.xml
-rw----     0.0 fat   116361 bx defN 21-Jul-24 22:20 repository.snapcase\fanart.jpg
-rw----     0.0 fat    12978 bx defN 21-Jul-24 22:20 repository.snapcase\icon.png
3 files, 130518 bytes uncompressed, 124659 bytes compressed:  4.5%
```


Applying [this patch ](https://gist.githubusercontent.com/asavah/f1dbd90a839fd77faffe4a354b618940/raw/45fbffdcf3b1e52ca0eab8f737aa4c9899859a54/create_repository_win.patch)fixed the issue for me.

**Comment by chadparry** [comment]
Thanks for the investigation!

So it looks like the `/` and `\` slash discrepancy was a red herring all along, right? Based on the Python documentation and previous bugs, it looks like it's best to use `os.sep` for the git APIs and writing ZIPs, and `/` when reading ZIPs. The tool handles that correctly with the `get_posix_path` helper. See https://github.com/chadparry/kodi-repository.chad.parry.org/commit/12ec20bbee18ae894969864fcc6fd75d8aa85650#diff-51c21e4f2c93dbc2659589670b3a621430ed128a8e9b3dbf83568a50cf19c0bb .

I can't apply this patch as-is, because it creates non-conforming ZIP archives that could break on POSIX machines. I need to write a different workaround that creates the directory entry and preserves forward slashes. Do you know which part of the patch will ensure that? If not, I'll find a Windows installation and experiment.

**Comment by chadparry** [comment]
I did some experiments, and the git relative paths need to be POSIX paths to fix this. I also verified that the linked patch only fixes this issue for subdirectories of git repositories. It doesn't fix the scenarios where the repository is a git root or a local folder. I'm preparing a fix now.

**Comment by chadparry** [comment]
I've released version 2.3.7. If you have a chance, please give it a try.

**Comment by zim514** [comment]
Thanks, looks like it works perfectly for me!
